Sources/TSCclibc/process.c

```diff
@@ -32,4 +36,61 @@ bool SPM_posix_spawn_file_actions_addchdir_np_supported() {
 #endif
 }
 
+int SPM_fork_exec_chdir(pid_t *pid, const char *cwd,
+                        const char *cmd, char *const argv[], char *const envp[],
+                        int in_pipe[], int out_pipe[], int err_pipe[], bool redirect_out, bool redirect_err) {
+    *pid = fork();
+
+    if (*pid < 0) {
+        perror("fork() failed");
+        _exit(EXIT_FAILURE);
+    } else if (*pid > 0) { // Parent process
+        // Wait for child process to finish
+        int status;
+        waitpid(*pid, &status, 0);
+        return status;
+    } else { // Child process
+        // Change working directory then execute cmd
+        if (chdir(cwd)) {
+            _exit(EXIT_FAILURE);
+        }
+        
+        // Replicate pipe logic in TSCBasic.Process.launch()
+        
+        // Dupe the read portion of the remote to 0.
+        dup2(in_pipe[0], 0);
+        // Close the other side's pipe since it was duped to 0.
+        close(in_pipe[0]);
+        close(in_pipe[1]);
+        
+        if (redirect_out) {
+            // Open the write end of the pipe.
+            dup2(out_pipe[1], 1);
+            // Close the other ends of the pipe since they were duped to 1.
+            close(out_pipe[0]);
+            close(out_pipe[1]);
+            
+            if (redirect_err) {
+                // If merged was requested, send stderr to stdout.
+                dup2(1, 2);
+            } else {
+                // If no redirect was requested, open the pipe for stderr.
+                dup2(err_pipe[1], 2);
+                // Close the other ends of the pipe since they were dupped to 2.
+                close(err_pipe[0]);
+                close(err_pipe[1]);
+            }
+        } else {
+            dup2(1, 1);
+            dup2(2, 2);
+        }
+
```

Comment:
there are a bunch of things missing here:

- we need a loop to close all other inherited file descriptors
- the signal mask is also inherited into the child. On dispatch worker threads pretty much all signals are blocked. So you'd inherit that which means that even a `kill CHILD_PID` would not kill the child. So you'd want to reset the signal mask 
- here's some code that I have written in a previous life: https://github.com/BromiumInc/BromiumCoreUtils/blob/2ec1c52a09831893618b9620383015e4362f7aa5/BromiumCoreUtils/BRUTask.m#L277-L347 . It's a mostly complete `NSTask` re-implementation in Obj-C. But the fork/exec code is pretty much completely straight C anyway, so you can look at that (it's BSD licensed)